### PR TITLE
Compress marks and primary key by default

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Compress marks and primary key by default. It significantly reduces the cold query time. Upgrade notes: the support for compressed marks and primary key has been added in version 22.9. If you turned on compressed marks or primary key or installed version 23.5 or newer, which has compressed marks or primary key on by default, you will not be able to downgrade to version 22.8 or earlier. You can also explicitly disable compressed marks or primary keys by specifying the `compress_marks` and `compress_primary_key` settings in the `<merge_tree>` section of the server configuration file. **Upgrade notes:** If you upgrade from versions prior to 22.9, you should either upgrade all replicas at once or disable the compression before upgrade, or upgrade through an intermediate version, where the compressed marks are supported but not enabled by default, such as 23.3.

### CODE REVIEW

**Review by SmitaRKulkarni — DISMISSED** [review]
LGTM

**Comment by alexey-milovidov** [comment]
@SmitaRKulkarni compressed marks are definitely not ready for production, for example, see 
```
01114_materialize_clear_index_compact_parts | FAIL | 0.10
2022-10-23 03:17:00 Received exception from server (version 22.10.1): 2022-10-23 03:17:00 Code: 341. DB::Exception: Received from localhost:9000. DB::Exception: Exception happened during execution of mutation 'mutation_3.txt' with part '2_2_2_0' reason: 'std::exception. Code: 1001, type: std::__1::__fs::filesystem::filesystem_error, e.what() = filesystem error: in file_size: No such file or directory ["/fasttest-workspace/db-fasttest/store/ec2/ec2b4711-5561-453a-bf23-575b196955a8/tmp_clone_2_2_2_0_3/primary.idx"]
```

**Comment by alexey-milovidov** [comment]
Postponed to version 22.3 to provide a window for upgrade/downgrade.

**Comment by alexey-milovidov** [comment]
Looks like everything is good. Let's reopen in four months.

**Comment by alexey-milovidov** [comment]
cat

**Comment by Algunenano** [comment]
I think this would need to be tagged as a breaking change as it needs manual intervention during an upgrade. If you have a cluster of replicas, when you update one of them the setting must be false or it will break the older replicas. Then when all of them are updated and now how to read compressed marks and primary key you can enable the setting.

**Comment by Algunenano** [comment]
I don't disagree with the change, just with how it's advertised. I don't expect seeing changes that break upgrades / running clusters outside the breaking change tag.

**Comment by alexey-milovidov** [comment]
Added a note for changelog.

**Comment by qoega** [comment]
Let's merge?

**Comment by alexey-milovidov** [comment]
It's not ready, one test has to be updated.

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit 55a519e692ef774a040ae56605bff9d6b5addfcd with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/42587/55a519e692ef774a040ae56605bff9d6b5addfcd/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Core/Settings.h [hunk 1]
```diff
@@ -733,7 +733,7 @@ class IColumn;
     M(Bool, multiple_joins_try_to_keep_original_names, false, "Do not add aliases to top level expression list on multiple joins rewrite", 0) \
     M(UInt64, grace_hash_join_initial_buckets, 1, "Initial number of grace hash join buckets", 0) \
     M(UInt64, grace_hash_join_max_buckets, 1024, "Limit on the number of grace hash join buckets", 0) \
-    M(Bool, optimize_distinct_in_order, true, "Enable DISTINCT optimization if some columns in DISTINCT form a prefix of sorting. For example, prefix of sorting key in merge tree or ORDER BY statement", 0) \
+    M(Bool, optimize_distinct_in_order, false, "Enable DISTINCT optimization if some columns in DISTINCT form a prefix of sorting. For example, prefix of sorting key in merge tree or ORDER BY statement", 0) \
```
Comment (azat) [inline src/Core/Settings.h:736]: @alexey-milovidov any clue why this has been disabled in together with enabling compressed marks by default?

**Comment by alexey-milovidov** [comment]
Yes, it was committed to git while switching the branches.
As this commit duplicates another commit in the master, it is normal.

**Pull request merged**